Samsung: Too many alarms (500) registered

Reported on Reddit

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/c956c3d3-f8aa-4406-82ac-cd79d848b596

```
java.lang.RuntimeException: Unable to create application com.ichi2.anki.AnkiDroidApp: java.lang.SecurityException: !@Too many alarms (500) registered from uid 10215
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6163)
at android.app.ActivityThread.access$1200(ActivityThread.java:235)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1768)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:6986)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445)
Caused by: java.lang.SecurityException: !@Too many alarms (500) registered from uid 10215
at android.os.Parcel.createException(Parcel.java:1966)
at android.os.Parcel.readException(Parcel.java:1934)
at android.os.Parcel.readException(Parcel.java:1884)
at android.app.IAlarmManager$Stub$Proxy.set(IAlarmManager.java:240)
at android.app.AlarmManager.setImpl(AlarmManager.java:722)
at android.app.AlarmManager.setRepeating(AlarmManager.java:468)
at com.ichi2.anki.services.BootService.scheduleDeckReminder(BootService.java:19)
at com.ichi2.anki.services.BootService.onReceive(BootService.java:8)
at com.ichi2.anki.AnkiDroidApp.onCreate(AnkiDroidApp.java:31)
at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1154)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6158)
... 8 more
```

https://stackoverflow.com/questions/29344971/java-lang-securityexception-too-many-alarms-500-registered-from-pid-10790-u

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
@@ -10,6 +10,8 @@ import android.preference.PreferenceManager;
 
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.Preferences;
+import com.ichi2.anki.R;
+import com.ichi2.anki.UIUtils;
 import com.ichi2.libanki.Collection;
 import com.ichi2.utils.Permissions;
 
@@ -17,6 +19,7 @@ import com.ichi2.utils.JSONObject;
 
 import java.util.Calendar;
 
+import androidx.annotation.NonNull;
 import timber.log.Timber;
 
 public class BootService extends BroadcastReceiver {
@@ -27,6 +30,7 @@ public class BootService extends BroadcastReceiver {
      */
     private static boolean sWasRun = false;
 
+    private boolean mFailedToShowNotifications = false;
 
     @Override
     public void onReceive(Context context, Intent intent) {
@@ -46,9 +50,29 @@ public class BootService extends BroadcastReceiver {
         }
 
         Timber.i("Executing Boot Service");
-        scheduleDeckReminder(context);
-        scheduleNotification(context);
+        catchAlarmManagerErrors(context, () -> scheduleDeckReminder(context));
+        catchAlarmManagerErrors(context, () -> scheduleNotification(context));
+        mFailedToShowNotifications = false;
         sWasRun = true;
+    }
+
+    private void catchAlarmManagerErrors(@NonNull Context context, @NonNull Runnable runnable) {
+        //#6332 - Too Many Alarms on Samsung Devices - this stops a fatal startup crash.
+        //We warn the user if they breach this limit
+        Integer error = null;
+        try {
+            runnable.run();
+        } catch (SecurityException ex) {
+            error = R.string.boot_service_too_many_notifications;
+        } catch (Exception e) {
+            error = R.string.boot_service_failed_to_schedule_notifications;
+        }
+        if (error != null) {
+            if (!mFailedToShowNotifications) {
+                UIUtils.showThemedToast(context, context.getString(error), false);
+            }
+            mFailedToShowNotifications = true;
+        }
     }
 
 

--- a/AnkiDroid/src/main/res/values/03-dialogs.xml
+++ b/AnkiDroid/src/main/res/values/03-dialogs.xml
@@ -193,4 +193,8 @@
 
     <!-- Browser Appearance -->
     <string name="card_template_editor_card_browser_appearance_failed">Could not load Card Browser Appearance</string>
+
+    <!-- Boot Service -->
+    <string name="boot_service_failed_to_schedule_notifications">Failed to schedule reminders</string>
+    <string name="boot_service_too_many_notifications">Too many reminders scheduled. Some will not appear</string>
 </resources>

